src/mimefactory.rs

```diff
@@ -744,58 +169,781 @@ impl<'a> MimeFactory<'a> {
+            // unless determined otherwise we add the Autocrypt header
+            let aheader = encrypt_helper.get_aheader().to_string();
+            unprotected_headers.push(Header::new("Autocrypt".into(), aheader));
+        }
+
+        protected_headers.push(Header::new("Subject".into(), subject));
+
+        let peerstates = self.peerstates_for_recipients()?;
+        let should_encrypt =
+            encrypt_helper.should_encrypt(self.context, e2ee_guranteed, &peerstates)?;
+        let is_encrypted = should_encrypt && force_plaintext == 0;
+
+        // Add gossip headers
+        if do_gossip {
+            info!(
+                self.context,
+                "gossiping headers for {} peerstates",
+                peerstates.len()
+            );
+            for peerstate in peerstates.iter().filter_map(|(state, _)| state.as_ref()) {
+                if peerstate.peek_key(min_verified).is_some() {
+                    if let Some(header) = peerstate.render_gossip_header(min_verified) {
+                        protected_headers.push(Header::new("Autocrypt-Gossip".into(), header));
+                    }
+                }
+            }
+        }
+
+        let rfc724_mid = match self.loaded {
+            Loaded::Message => self.msg.rfc724_mid.clone(),
+            Loaded::MDN => dc_create_outgoing_rfc724_mid(None, &self.from_addr),
+        };
+
+        protected_headers.push(Header::new("Message-ID".into(), rfc724_mid.clone()));
+
+        unprotected_headers.push(Header::new_with_value("To".into(), to).unwrap());
+        unprotected_headers.push(Header::new_with_value("From".into(), vec![from]).unwrap());
+
+        let outer_message = if is_encrypted {
+            // Store protected headers in the inner message.
+            for header in protected_headers.into_iter() {
+                message = message.header(header);
+            }
+            // Set the appropriate Content-Type for the inner message.
+            let mut existing_ct = message
+                .get_header("Content-Type".to_string())
+                .and_then(|h| h.get_value::<String>().ok())
+                .unwrap_or_else(|| "text/plain; charset=utf-8;".to_string());
+
+            if !existing_ct.ends_with(';') {
+                existing_ct += ";";
+            }
+            message = message.replace_header(Header::new(
+                "Content-Type".to_string(),
+                format!("{} protected-headers=\"v1\";", existing_ct),
+            ));
+
+            // Set the appropriate Content-Type for the outer message
+            let mut outer_message = PartBuilder::new().header((
+                "Content-Type".to_string(),
+                "multipart/encrypted; protocol=\"application/pgp-encrypted\";".to_string(),
+            ));
+
+            // Store the unprotected headers on the outer message.
+            for header in unprotected_headers.into_iter() {
+                outer_message = outer_message.header(header);
+            }
```

Comment:
Is this correct? It looks as if it overwrites outer_message all the time, instead of accumulating headers.

Reply:
ah -- it's a builder pattern! now i get it